java-squid/src/main/java/org/sonar/java/resolve/JavaType.java

```diff
@@ -384,5 +396,111 @@ public JavaType substitution(TypeVariableJavaType typeVariableType) {
       }
       return Lists.newArrayList();
     }
+
+    @Override
+    public boolean isSubtypeOf(Type superType) {
+      if (erasure().isSubtypeOf(superType.erasure())) {
+        boolean superTypeIsParametrizedJavaType = superType instanceof ParametrizedTypeJavaType;
+        if (superTypeIsParametrizedJavaType) {
+          return checkSubstitutedTypesCompatibility((ParametrizedTypeJavaType) superType);
+        }
+        return !superTypeIsParametrizedJavaType;
+      }
+      return false;
+    }
+
+    @Override
+    public boolean usesWildCard() {
+      for (JavaType substitution : typeSubstitution.substitutedTypes()) {
+        if (substitution.isTagged(WILDCARD)) {
+          return true;
+        }
+      }
+      return false;
+    }
+
+    private boolean checkSubstitutedTypesCompatibility(ParametrizedTypeJavaType superType) {
+      List<JavaType> myTypes = typeSubstitution.substitutedTypes();
+      List<JavaType> itsTypes = superType.typeSubstitution.substitutedTypes();
+      if (itsTypes.size() != myTypes.size()) {
+        return false;
+      }
+      for (int i = 0; i < myTypes.size(); i++) {
+        JavaType myType = myTypes.get(i);
+        JavaType itsType = itsTypes.get(i);
+        if (itsType.isTagged(WILDCARD)) {
+          if (!myType.isSubtypeOf(itsType)) {
+            return false;
+          }
+        } else if (!myType.equals(itsType)) {
+          return false;
+        }
+      }
+      return true;
+    }
+  }
+
+  public static class WildCardType extends JavaType {
+
+    public enum BoundType {
+      UNBOUNDED("?"),
+      SUPER("? super "),
+      EXTENDS("? extends ");
+
+      private final String name;
+
+      BoundType(String name) {
+        this.name = name;
+      }
+
+      @Override
+      public String toString() {
+        return name;
+      }
+    }
+
+    private final JavaType bound;
+    private final BoundType boundType;
+
+    public WildCardType(JavaType bound, BoundType boundType) {
+      super(WILDCARD, new JavaSymbol.WildcardSymbol(boundType == BoundType.UNBOUNDED ? boundType.toString() : (boundType + bound.symbol().name())));
+      this.bound = bound;
+      this.boundType = boundType;
+      this.symbol.type = this;
+    }
+
+    public boolean isUnbounded() {
+      return boundType == BoundType.UNBOUNDED;
+    }
+
+    @Override
+    public boolean isSubtypeOf(String fullyQualifiedName) {
+      return false;
+    }
+
+    @Override
+    public boolean isSubtypeOf(Type superType) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "isSubtypeOf" is 11 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)